File: src/fabricflow/fibc/lib/fibccfg.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright (C) 2017 Nippon Telegraph and Telephone Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
FIB Controller configuration
"""

_KEY_ROUTERS = "routers"
_KEY_DPATHS = "datapaths"
_CFG_PATTERN = "fibc*.yml"

class Config(dict):
    """
    Port config
    """

    def __init__(self, *args, **kwargs):
        """
        routers:
        - re_id   : <string>
          datapath: <string>
          ports   :
          - { name: <string>, port: <int> }

        datapaths:
        - name  : <string>
          dp_id : <int>
          mode  : "geneic" or "ofdpa2" or "ovs"
        """

        super(Config, self).__init__(*args, **kwargs)
        self.routers = list()
        self.dpaths = list()


    def load(self, stream):
        """
        Load config and append to fields.
        """
        import yaml
        datas = yaml.load(stream)
        self.routers.extend(datas.get(_KEY_ROUTERS, []))
        self.dpaths.extend(datas.get(_KEY_DPATHS, []))
        return self


    def get_datapath(self, name):
        """
        get datapath entry.
        """
        for dpath in self.dpaths:
            if dpath["name"] == name:
                return dpath

        return None


    def get_router(self, re_id):
        """
        get router entry
        """
        for router in self.routers:
            if router["re_id"] == re_id:
                return router

        return None


    def extend_port(self, router, port):
        """
        append router, datapath info to path
        """
        dpath = self.get_datapath(router["datapath"])
        if dpath is None:
            return None

        ext_port = dict(**port)
        ext_port["re_id"] = router["re_id"]
        ext_port["dp_id"] = dpath["dp_id"]
        ext_port["mode"] = dpath["mode"]

        return ext_port


    def extend_router(self, router):
        """
        append datapath info to router
        """
        dpath = self.get_datapath(router["datapath"])
        if dpath is None:
            return None

        ext_ports = list()
        for port in router["ports"]:
            ext_port = self.extend_port(router, port)
            if ext_port is not None:
                ext_ports.append(ext_port)

        ext_router = dict(**router)
        ext_router["dp_id"] = dpath["dp_id"]
        ext_router["mode"] = dpath["mode"]
        ext_router["ports"] = ext_ports

        return ext_router


    def get_ext_routers(self):
        """
        get extended routers
        """
        ext_routers = list()
        for router in self.routers:
            ext_router = self.extend_router(router)
            if ext_router is not None:
                ext_routers.append(ext_router)

        return ext_routers


def load_dir(dirpath):
    """
    Load config files in directory.
    """
    import glob
    import os.path
    cfg = Config()
    for path in glob.glob(os.path.join(dirpath, _CFG_PATTERN)):
        with open(path, "r") as stream:
            cfg.load(stream)
    return cfg


def print_router(router):
    """
    Print router
    """
    print "Desc: '{desc}', REID: '{re_id}', DP: '{datapath}'".format(**router)
    for port in router["ports"]:
        print "  Port: {0}".format(port)


def print_datapath(dpath):
    """
    Print datapath.
    """
    print "Name: '{name}', DpId: {dp_id}/0x{dp_id:x}, Mode: '{mode}'".format(**dpath)


def _main():
    cfg = load_dir(".")

    for router in cfg.routers:
        print_router(router)

    for dpath in cfg.dpaths:
        print_datapath(dpath)


if __name__ == "__main__":
    _main()

File: src/fabricflow/fibc/cmd/lib/fibccfg.go
// -*- coding: utf-8 -*-

// Copyright (C) 2018 Nippon Telegraph and Telephone Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package fibccmd

import (
	"fmt"
)

const (
	CFGCMD_ADD      = "add"
	CFGCMD_DEL      = "delete"
	CFGTBL_PORT     = "port"
	CFGTBL_DPID     = "dp"
	CFGTBL_REID     = "re"
	CFGMODE_GENERIC = "generic"
	CFGMODE_OFDPA2  = "ofdpa2"
	CFGMODE_OVS     = "ovs"
)

type Port struct {
	Name   string   `yaml:"name" json:"name"`
	Port   uint32   `yaml:"port" json:"port"`
	Link   string   `yaml:"link" json:"link"`
	Slaves []string `yaml:"slaves" json:"slaves"`
}

func NewPort(port uint32, name string) *Port {
	return &Port{
		Name:   name,
		Port:   port,
		Link:   "",
		Slaves: []string{},
	}
}

func NewVLANPort(port uint32, name string, vid uint16) *Port {
	return &Port{
		Name:   fmt.Sprintf("%s.%d", name, vid),
		Port:   port,
		Link:   name,
		Slaves: []string{},
	}
}

func (p *Port) String() string {
	return fmt.Sprintf("name:'%s', port:%d, link='%s', slaves=%v", p.Name, p.Port, p.Link, p.Slaves)
}

type Router struct {
	Desc     string  `yaml:"desc" json:"desc"`
	ReId     string  `yaml:"re_id" json:"re_id"`
	Datapath string  `yaml:"datapath" json:"datapath"`
	Ports    []*Port `yaml:"ports" json:"ports"`
}

func NewRouter(reid string, dpname string) *Router {
	return &Router{
		ReId:     reid,
		Datapath: dpname,
		Ports:    []*Port{},
	}
}

func (r *Router) AddPort(ports ...*Port) {
	r.Ports = append(r.Ports, ports...)
}

func (r *Router) String() string {
	return fmt.Sprintf("ReId:'%s', Dpath:'%s'", r.ReId, r.Datapath)
}

type Datapath struct {
	Name string `yaml:"name" json:"name"`
	Dpid uint64 `yaml:"dp_id" json:"dp_id"`
	Mode string `yaml:"mode" json:"mode"`
}

func NewDatapath(name string, dpid uint64, mode string) *Datapath {
	return &Datapath{
		Name: name,
		Dpid: dpid,
		Mode: mode,
	}
}

func (d *Datapath) String() string {
	return fmt.Sprintf("name:'%s', dpid:%d, mode:'%s'", d.Name, d.Dpid, d.Mode)
}

type Config struct {
	Routers   []*Router   `yaml:"routers" json:"routers"`
	Datapaths []*Datapath `yaml:"datapaths" json:"datapaths"`
}

type ConfigClient struct {
	url string
}

func NewConfigClient(addr string) *ConfigClient {
	return &ConfigClient{
		url: fmt.Sprintf("http://%s/fib/portmap", addr),
	}
}

func (m *ConfigClient) Add(cfg *Config) error {
	if err := m.ModDpaths(CFGCMD_ADD, cfg.Datapaths); err != nil {
		return err
	}

	if err := m.ModRouters(CFGCMD_ADD, cfg.Routers); err != nil {
		return err
	}

	if err := m.ModPorts(CFGCMD_ADD, cfg.Routers); err != nil {
		return err
	}

	return nil
}

func (m *ConfigClient) Del(cfg *Config) error {
	if err := m.ModPorts(CFGCMD_DEL, cfg.Routers); err != nil {
		return err
	}

	if err := m.ModRouters(CFGCMD_DEL, cfg.Routers); err != nil {
		return err
	}

	if err := m.ModDpaths(CFGCMD_DEL, cfg.Datapaths); err != nil {
		return err
	}

	return nil
}

func (m *ConfigClient) ModDpaths(cmd string, dpaths []*Datapath) error {
	url := fmt.Sprintf("%s/%s/%s", m.url, CFGTBL_DPID, cmd)
	for _, dpath := range dpaths {
		if err := httpJson(url, dpath); err != nil {
			return err
		}
	}
	return nil
}

func (m *ConfigClient) ModRouters(cmd string, routers []*Router) error {
	url := fmt.Sprintf("%s/%s/%s", m.url, CFGTBL_REID, cmd)
	for _, router := range routers {
		if err := httpJson(url, router); err != nil {
			return err
		}
	}
	return nil
}

func (m *ConfigClient) ModPorts(cmd string, routers []*Router) error {
	url := fmt.Sprintf("%s/%s/%s", m.url, CFGTBL_PORT, cmd)
	for _, router := range routers {
		if err := httpJson(url, router); err != nil {
			return err
		}
	}
	return nil
}
